File: Python/A_Project/C_Elevator_System/UtilityClasses/elevator.py
from collections import deque

from CommonEnums.direction import Direction
from CommonEnums.elevator_state import ElevatorState


class Elevator:
    def __init__(self, elevator_id: int):
        self._id = elevator_id
        self._current_floor = 0
        self._direction = Direction.IDLE
        self._state = ElevatorState.IDLE
        self._observers = []
        self._requests = deque()

    def add_observer(self, observer):
        self._observers.append(observer)

    def remove_observer(self, observer):
        if observer in self._observers:
            self._observers.remove(observer)

    def notify_state_change(self):
        for observer in self._observers:
            observer.on_elevator_state_change(self, self._state)

    def notify_floor_change(self, floor: int):
        for observer in self._observers:
            observer.on_elevator_floor_change(self, floor)

    def get_direction(self) -> Direction:
        return self._direction

    def set_direction(self, new_direction: Direction):
        self._direction = new_direction

    def set_state(self, new_state: ElevatorState):
        self._state = new_state
        self.notify_state_change()

    def add_request(self, request):
        if request not in self._requests:
            self._requests.append(request)

        requested_floor = request.get_floor()
        if self._state == ElevatorState.IDLE and request is not None:
            if requested_floor == self._current_floor:
                self._complete_arrival()
                return
            if requested_floor > self._current_floor:
                self.set_direction(Direction.UP)
            elif requested_floor < self._current_floor:
                self.set_direction(Direction.DOWN)
            self.set_state(ElevatorState.MOVING)

    def move_to_next_stop(self, next_stop: int):
        if self._state != ElevatorState.MOVING:
            return

        if next_stop == self._current_floor:
            self._complete_arrival()
            return

        self.set_direction(
            Direction.UP if next_stop > self._current_floor else Direction.DOWN
        )

        while self._current_floor != next_stop:
            self._current_floor += 1 if self._direction == Direction.UP else -1
            self.notify_floor_change(self._current_floor)
        self._complete_arrival()

    def _complete_arrival(self):
        self.set_state(ElevatorState.STOPPED)
        self._requests = deque(
            [request for request in self._requests if request.get_floor() != self._current_floor]
        )
        if not self._requests:
            self.set_direction(Direction.IDLE)
            self.set_state(ElevatorState.IDLE)
        else:
            self.set_state(ElevatorState.MOVING)

    def complete_request(self):
        self._complete_arrival()

    def get_elevator_id(self) -> int:
        return self._id

    def get_current_floor(self) -> int:
        return self._current_floor

    def get_state(self) -> ElevatorState:
        return self._state

    def get_request_queue(self):
        return list(self._requests)

    def get_destination_floors(self):
        return list(self._requests)

File: Ruby/A_Project/C_Elevator_System/UtilityClass/elevator.rb
require_relative '../CommonEnums/direction'
require_relative '../CommonEnums/elevator_state'

class Elevator
  def initialize(elevator_id)
    @id            = elevator_id
    @current_floor = 0
    @direction     = Direction::IDLE
    @state         = ElevatorState::IDLE
    @observers     = []
    @requests      = []
  end

  # ── Observer management ────────────────────────────────────────────────────

  def add_observer(observer)
    @observers << observer
  end

  def remove_observer(observer)
    @observers.delete(observer)
  end

  def notify_state_change
    @observers.each { |obs| obs.on_elevator_state_change(self, @state) }
  end

  def notify_floor_change(floor)
    @observers.each { |obs| obs.on_elevator_floor_change(self, floor) }
  end

  # ── Direction / State ──────────────────────────────────────────────────────

  def get_direction
    @direction
  end

  def set_direction(new_direction)
    @direction = new_direction
  end

  def get_state
    @state
  end

  def set_state(new_state)
    @state = new_state
    notify_state_change
  end

  # ── Request management ─────────────────────────────────────────────────────

  def add_request(request)
    @requests << request unless @requests.include?(request)

    requested_floor = request.get_floor

    if @state == ElevatorState::IDLE
      if requested_floor == @current_floor
        complete_arrival
        return
      end

      set_direction(requested_floor > @current_floor ? Direction::UP : Direction::DOWN)
      set_state(ElevatorState::MOVING)
    end
  end

  def move_to_next_stop(next_stop)
    return unless @state == ElevatorState::MOVING

    if next_stop == @current_floor
      complete_arrival
      return
    end

    set_direction(next_stop > @current_floor ? Direction::UP : Direction::DOWN)

    until @current_floor == next_stop
      @current_floor += (@direction == Direction::UP ? 1 : -1)
      notify_floor_change(@current_floor)
    end

    complete_arrival
  end

  def complete_request
    complete_arrival
  end

  # ── Accessors ──────────────────────────────────────────────────────────────

  def get_elevator_id
    @id
  end

  def get_current_floor
    @current_floor
  end

  def get_request_queue
    @requests.dup
  end

  def get_destination_floors
    @requests.dup
  end

  private

  def complete_arrival
    set_state(ElevatorState::STOPPED)
    @requests.reject! { |req| req.get_floor == @current_floor }

    if @requests.empty?
      set_direction(Direction::IDLE)
      set_state(ElevatorState::IDLE)
    else
      set_state(ElevatorState::MOVING)
    end
  end
end
